### PR TITLE
[#363] refactor: Make the coordinator client managed by CoordinatorClientFactory singleton

### PR DESCRIPTION
**Pull request by cchung100m**
Hi @zuston @jerqi @xianjingfeng 

I am re-submitting a PR to prevent potential conflict since the last PR is been a while, any comments and suggestions would be appreciated if you are available, thank you.

### What changes were proposed in this pull request?

Make the coordinator client managed by CoordinatorClientFactory singleton

### Why are the changes needed?

Fix: #363

### Does this PR introduce _any_ user-facing change?

No.

### How was this patch tested?

current UT

### CODE REVIEW

File: internal-client/src/main/java/org/apache/uniffle/client/factory/CoordinatorClientFactory.java [hunk 1]
```diff
@@ -29,47 +30,68 @@
...
     String[] coordinatorList = coordinators.trim().split(",");
     if (coordinatorList.length == 0) {
       String msg = "Invalid " + coordinators;
       LOG.error(msg);
       throw new RssException(msg);
     }
-
+    List<CoordinatorClient> coordinatorClients = Lists.newLinkedList();
     for (String coordinator : coordinatorList) {
+      clients.computeIfAbsent(type, key -> JavaUtils.newConcurrentMap());
```
Comment (xianjingfeng) [inline internal-client/src/main/java/org/apache/uniffle/client/factory/CoordinatorClientFactory.java:77]: coordinatorClients.add(getOrCreateCoordinatorClient(type, coordinator))?
I think `getOrCreateCoordinatorClients` can be more simple.

File: internal-client/src/main/java/org/apache/uniffle/client/factory/CoordinatorClientFactory.java [hunk 2]
```diff
@@ -29,47 +30,68 @@
...
+      CoordinatorClient coordinatorClient = typeToClients.get(coordinator);
+      if (coordinatorClient != null) {
+        return coordinatorClient;
+      }
+    }
+
+    switch (ClientType.valueOf(type)) {
+      case GRPC:
+        CoordinatorClient grpcClient = new CoordinatorGrpcClient(host, port);
+        typeToClients.put(coordinator, grpcClient);
```
Comment (xianjingfeng) [inline internal-client/src/main/java/org/apache/uniffle/client/factory/CoordinatorClientFactory.java:60]: Would `typeToClients` be null?

Comment (cchung100m) [reply internal-client/src/main/java/org/apache/uniffle/client/factory/CoordinatorClientFactory.java:60]: Good catch! it should be null here.

File: internal-client/src/main/java/org/apache/uniffle/client/factory/CoordinatorClientFactory.java (hunk 1 above)
Comment (cchung100m) [reply internal-client/src/main/java/org/apache/uniffle/client/factory/CoordinatorClientFactory.java]: Yes, we should leave the work to `getOrCreateCoordinatorClients`

**Comment by cchung100m** [comment]
Hi @jerqi 
I am still working on the integration test failure. May I know what is this test actually doing?

https://github.com/apache/incubator-uniffle/blob/13c71438c5fd614cdd08939d160afa722475beee/integration-test/common/src/test/java/org/apache/uniffle/test/CoordinatorAssignmentTest.java#L163

Do we expect the server to change the shuffle nodes to `5`?
https://github.com/apache/incubator-uniffle/blob/13c71438c5fd614cdd08939d160afa722475beee/integration-test/common/src/test/java/org/apache/uniffle/test/CoordinatorAssignmentTest.java#L178

I am not sure why the `info.getServerToPartitionRanges().keySet().size()` returns `10` here
https://github.com/apache/incubator-uniffle/blob/13c71438c5fd614cdd08939d160afa722475beee/integration-test/common/src/test/java/org/apache/uniffle/test/CoordinatorAssignmentTest.java#L182

**Pull request closed without merging**